File: 1011 Capacity To Ship Packages Within D Days/capacity_to_ship_packages_within_d_days.cpp
/*
BY RUSHABH GALA
rushabhvg
GitHub :
https://github.com/rushabhvg
StackOverflow :
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

class Solution {
    int maxSum=INT_MAX;
public:
    int shipWithinDays(vector<int>& weights, int days) {
        int ans=0;
        int low = *max_element(weights.begin(), weights.end());
        int high = accumulate(weights.begin(), weights.end(), 0);
        while(low<=high) {
            int mid = low + (high-low)/2;
            int cnt = 1;
            int sum = 0;
            for (int w : weights) {
                sum += w;
                if (sum > mid) {
                    cnt++;
                    sum = w;
                }
            }
            if (cnt <= days) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }
};

/*
BY RUSHABH GALA
rushabhvg
GitHub :
https://github.com/rushabhvg
StackOverflow :
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.java
/*
By Rushabh Gala
rushabhvg
GitHub:
https://github.com/rushabhvg
StackOverflow:
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/

class Solution {
    public int shipWithinDays(int[] weights, int days) {
        int n = weights.length;
        int st = 1;
        int end = 0;
        int mid = 0;
        for(int w:weights) {
            if(w>st) {
                st = w;
            }
            end += w;
        }
        while(st<end) {
            mid = (st + end)/2;
            int d = 1;
            int tot = 0;
            for(int w:weights) {
                tot += w;
                if(tot>mid) {
                    tot = w;
                    d += 1;
                }
                if(d>days) break;
            }
            if(d<=days) {
                end = mid;
            } else {
                st = mid+1;
            }
        }
        return st;
    }
}

/*
By Rushabh Gala
rushabhvg
GitHub:
https://github.com/rushabhvg
StackOverflow:
https://stackoverflow.com/users/16571212/rushabhvg
LinkedIn:
https://www.linkedin.com/in/rushabhvg/
*/
